e2e/partition_planner.go

```diff
@@ -0,0 +1,938 @@
+		return nil, fmt.Errorf("replication factor %d exceeds available brokers %d", p.cfg.ReplicationFactor, len(meta.Brokers))
+	}
+
+	// Build state required for the planning
+	state := BuildState(meta)
+	desired := ComputeDesired(state, p.cfg)
+	tracker := NewLoadTracker(state)
+	selector := NewRackAwareSelector(state, tracker)
+	p.sel = selector
+
+	b := NewPlanBuilder(state, desired, tracker)
+
+	// Phase 1: normalize RF and racks (low movement first)
+	// Grow/trim replicas to configured RF and re-pick to maximize unique racks
+	// per partition. We avoid moving the leader when possible.
+	fixReplicationAndRack(b, selector, p.cfg.ReplicationFactor)
+
+	// Phase 2: ensure each broker is preferred leader for >= 1 partition
+	// This guarantees probe coverage: if a broker dies, some partition leader is
+	// unavailable and the probe trips.
+	ensureLeaderCoverage(b, selector)
+
+	// Phase 3: ensure desired partition count
+	// Enforce >= max(current, #brokers, partitionsPerBroker * #brokers).
+	ensurePartitionCount(b, selector)
+
+	// Build final plan and log the changes
+	plan := b.Build()
+
+	// Log with appropriate level based on whether changes are needed
+	totalChanges := len(plan.Reassignments) + len(plan.CreateAssignments)
+	if totalChanges == 0 {
+		p.logger.Info("e2e probe topic partition leadership and replica distribution check completed - optimal",
+			zap.String("topic", state.TopicName),
+			zap.Int("brokers", len(state.Brokers)),
+			zap.Int("partitions", len(state.Partitions)),
+			zap.String("status", "optimal"),
+		)
+	} else {
+		p.logger.Info("plan to change partition leadership and replica placements on e2e topic has been prepared",
+			zap.String("topic", state.TopicName),
+			zap.Int("brokers", len(state.Brokers)),
+			zap.Int("cur_partitions", len(state.Partitions)),
+			zap.Int("final_partitions", plan.FinalPartitionCount),
+			zap.Int("reassignments", len(plan.Reassignments)),
+			zap.Int("creates", len(plan.CreateAssignments)),
+			zap.Int("total_changes", totalChanges),
+		)
+	}
+	return plan, nil
+}
+
+// -----------------------------------------------------------------------------
+// Planning phases
+// -----------------------------------------------------------------------------
+
+// PlanBuilder holds a predictive view of partition -> replicas after applying
+// staged operations. We never mutate ClusterState.Partitions; instead, we write
+// new assignments into PlanBuilder.view and record high-level operations to
+// produce Kafka requests at the end (see Plan.ToRequests).
+//
+// This keeps the planning phases simple and side-effect free.
+type PlanBuilder struct {
+	state   ClusterState
+	desired Desired
+	tracker *LoadTracker
+
+	// view is our predictive map: partitionID -> replicas (preferred leader at idx 0)
+	view map[int32][]int32
+
+	reassign []Reassignment     // staged reassignments for existing partitions
+	creates  []CreateAssignment // staged creations of new partitions
```

Comment:
nit/suggestion: I'd prefer both struct fields to either use singular or plural form to stay consistent, but not mixed

Reply:
Fixed!